pkg/controller/utils/datadog/forwarders_manager.go

```diff
@@ -0,0 +1,80 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package datadog
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ForwardersManager is a collection of metricsForwarder per DatadogAgentDeployment
+// ForwardersManager implements the controller-runtime Runnable interface
+type ForwardersManager struct {
+	forwarders       map[string]*metricsForwarder
+	k8sClient        client.Client
+	stoppedForwarder chan string
+	sync.Mutex
+}
+
+// NewForwardersManager builds a new ForwardersManager
+// ForwardersManager implements the controller-runtime Runnable interface
+func NewForwardersManager(k8sClient client.Client) *ForwardersManager {
+	return &ForwardersManager{
+		k8sClient:        k8sClient,
+		forwarders:       make(map[string]*metricsForwarder),
+		stoppedForwarder: make(chan string),
+	}
+}
+
+// Start must be handled by the controller-runtime manager
+func (f *ForwardersManager) Start(stop <-chan struct{}) error {
+	for {
+		select {
+		case <-stop:
+			f.stopAllForwarders()
+			return nil
+		case id := <-f.stoppedForwarder:
+			f.unregisterForwarder(id)
+		}
+	}
+}
+
+// Register starts a new metricsForwarder if a new DatadogAgentDeployment is detected
+func (f *ForwardersManager) Register(namespacedName types.NamespacedName) {
+	f.Lock()
+	defer f.Unlock()
+	id := namespacedName.String()
+	if _, found := f.forwarders[id]; !found {
+		log.Info("New Datadog metrics forwarder registred", "ID", id)
+		f.forwarders[id] = newMetricsForwarder(f.k8sClient, namespacedName)
+		go f.forwarders[id].start(f.stoppedForwarder)
+	}
+}
+
+// stopAllForwarders stops the running metricsForwarder goroutines
+func (f *ForwardersManager) stopAllForwarders() {
+	f.Lock()
+	defer f.Unlock()
+	for id, forwarder := range f.forwarders {
+		log.Info("Stopping metrics forwarder", "ID", id)
+		forwarder.stop <- struct{}{}
+	}
+}
+
+// unregisterForwarder deletes a given metricsForwarder
+func (f *ForwardersManager) unregisterForwarder(id string) {
+	f.Lock()
+	defer f.Unlock()
+	if _, found := f.forwarders[id]; !found {
+		log.Error(fmt.Errorf("%s not found", id), "cannot unregister metrics forwarder")
+		return
+	}
+	log.Info("Metrics forwarder unregistered", "ID", id)
+	delete(f.forwarders, id)
```

Comment:
I think you should stop it before the `delete`

Reply:
I don't think so, this is only called when the forwarder sends its ID to the manager and breaks it's infinite loop (when its `DatadogAgentDeployment` is deleted)

in the forwarder
```golang
					// DatadogAgentDeployment was deleted
					// inform the ForwardersManager before shutting down the metricsForwarder
					managerChan <- mf.id
					return
````

in the manager 
```golang
		case id := <-f.stoppedForwarder:
			f.unregisterForwarder(id)
```